File: ts/project/.sdk/src/cmp/go/fragment/EntityCreateOp.fragment.go
package entity

import (
	"GOMODULE/core"

	vs "github.com/voxgig/struct"
)

type entityCreateOp struct{}

// EJECT-START

func (e *EntyClass) Create(reqdata map[string]any, ctrl map[string]any) (any, error) {
	utility := e.utility
	ctx := utility.MakeContext(map[string]any{
		"opname":  "create",
		"ctrl":    ctrl,
		"match":   e.match,
		"data":    e.data,
		"reqdata": reqdata,
	}, e.entctx)

	return e.runOp(ctx, func() {
		if ctx.Result != nil {
			if ctx.Result.Resdata != nil {
				e.data = core.ToMapAny(vs.Clone(ctx.Result.Resdata))
				if e.data == nil {
					e.data = map[string]any{}
				}
			}
		}
	})
}

// CreateTyped is the statically-typed variant of Create: it takes an
// EntityNameCreateData and returns an EntityName. It delegates to the untyped
// Create (identical runtime) and converts at the typed boundary.
func (e *EntyClass) CreateTyped(reqdata EntityNameCreateData, ctrl map[string]any) (EntityName, error) {
	res, err := e.Create(asMap(reqdata), ctrl)
	if err != nil {
		return EntityName{}, err
	}
	return typedFrom[EntityName](res), nil
}

// EJECT-END

File: ts/project/.sdk/src/cmp/cpp/fragment/EntityCreateOp.fragment.cpp
// EJECT-START

  SdkEntityPtr create(const Value& reqdata, const Value& ctrl) override {
    CtxSpec cs;
    cs.setOpname("create");
    cs.ctrlMap = ctrl.is_map() ? ctrl : vmap();
    cs.match = this->match_;
    cs.data = this->data_;
    cs.reqdata = reqdata.is_map() ? reqdata : vmap();
    CtxPtr ctx = this->utility->makeContext(cs, this->entctx);

    runOp(ctx, [this, ctx]() {
      if (ctx->result) {
        if (!is_nullish(ctx->result->resdata)) {
          Value d = Helpers::toMapAny(Struct::clone(ctx->result->resdata));
          this->data_ = d.is_map() ? d : vmap();
        }
      }
    });

    // The operation resolves to THIS entity: runOp has just absorbed the
    // result into it, and the caller reaches the record through data().
    // See AGENTS.md "Entity operations return ENTITIES".

    return this->self();
  }

// EJECT-END
